### PR TITLE
Create raspbian install script

### PR DESCRIPTION
**Pull request by MarianRaphael**
## Description

Creation of a Raspbian install script

## Checklist

 - [ ] I have read the [contribution guidelines](https://github.com/flowforge/flowforge/blob/main/CONTRIBUTING.md)
 - [ ] Suitable unit/system level tests have been added and they pass 
 - [ ] Documentation has been updated
    - [ ] Upgrade instructions
    - [ ] Configuration details
    - [ ] Concepts
 - [ ] Changes `flowforge.yml`?
    - [ ] Issue/PR raised on `flowforge/helm` to update ConfigMap Template
    - [ ] Issue/PR raised on `flowforge/CloudProject` to update values for Staging/Production

## Labels

 - [ ] Backport needed? -> add the `backport` label
 - [ ] Includes a DB migration? -> add the `area:migration` label

### CODE REVIEW

**Comment by MarianRaphael** [comment]
Is script the right folder to place an installation helper script in?
Do we want such a script in the repo? Otherwise, I would just include it in docs?

File: service/raspbian-install-device-agent.sh [hunk 1]
```diff
@@ -0,0 +1,109 @@
+#!/bin/bash
+
+if [[ $EUID -ne 0 ]]; then
+    echo "This script must be run as root"
```
Comment (hardillb) [inline service/raspbian-install-device-agent.sh:4]: The flowfuse install script uses sudo for any actions that need root. On raspbian this works without needing a password, but the script will ask for password if needed

https://github.com/flowforge/installer/blob/a7909d1074106fa3f12abf3c8a4241eb6dcda44b/install.sh#L6

File: service/raspbian-install-device-agent.sh [hunk 2]
```diff
@@ -0,0 +1,109 @@
...
+    else
+        echo "**************************************************************"
+        echo " You need NodeJS $MIN_NODEJS or newer, please upgrade "
+        echo "**************************************************************"
+        exit 1
+    fi
+else
+    echo "**************************************************************"
+    echo " No NodeJS found"
+    echo " Do you want to install NodeJS $MIN_NODEJS?"
```
Comment (hardillb) [inline service/raspbian-install-device-agent.sh:36]: This should probably install at least NodeJS 18 if installing from new

File: service/raspbian-install-device-agent.sh [hunk 3]
```diff
@@ -0,0 +1,109 @@
+#!/bin/bash
+
+if [[ $EUID -ne 0 ]]; then
+    echo "This script must be run as root"
+    exit 1
+fi
+
+# Node.js version
+MIN_NODEJS=16
```
Comment (hardillb) [inline service/raspbian-install-device-agent.sh:9]: We technically support NodeJS 14 for the device agent

File: service/raspbian-install-device-agent.sh [hunk 4]
```diff
@@ -0,0 +1,109 @@
...
+        sudo apt-get update
+        sudo apt-get install nodejs -y
+    else
+        echo "**************************************************************"
+        echo " You will need to manually install NodeJS first. "
+        echo " Exiting."
+        echo "**************************************************************"
+        exit 1
+    fi
+fi
```
Comment (hardillb) [inline service/raspbian-install-device-agent.sh:58]: This won't work on ARMv6 builds as the standard NodeJS builds don't support it. We should document that Pi Zero's are not supported

File: service/raspbian-install-device-agent.sh [hunk 5]
```diff
@@ -0,0 +1,109 @@
...
+[Install]
+WantedBy=multi-user.target" | sudo tee /etc/systemd/system/flowforge-device-agent.service >/dev/null
+
+# Reload systemd, enable and start the service
+sudo systemctl daemon-reload
+sudo systemctl enable flowforge-device-agent.service
+sudo systemctl start flowforge-device-agent.service
+
+# Output status of the service
+sudo systemctl status flowforge-device-agent.service
```
Comment (hardillb) [inline service/raspbian-install-device-agent.sh:109]: We should either run all commands with sudo or none (assume the whole script is running as root)

**Review by hardillb — COMMENTED** [review]
Looks good, just a few comments

File: service/raspbian-install-device-agent.sh (hunk 3 above)
Comment (MarianRaphael) [reply service/raspbian-install-device-agent.sh:9]:
```suggestion
MIN_NODEJS=14
```

File: service/raspbian-install-device-agent.sh (hunk 2 above)
Comment (MarianRaphael) [reply service/raspbian-install-device-agent.sh:36]:
```suggestion
    echo " Do you want to install NodeJS 18?"
```

File: service/raspbian-install-device-agent.sh [hunk 6]
```diff
@@ -0,0 +1,109 @@
...
+    [ -z "$yn" ] && yn="n"
+
+    if [[ "$yn" == "y" ]] || [[ "$yn" == "Y" ]]; then
+        if [ ! -x "$(command -v curl)" ]; then
+            sudo apt-get install -y curl
+        fi
+        sudo apt-get install -y ca-certificates curl gnupg
+        sudo mkdir -p /etc/apt/keyrings
+        curl -fsSL https://deb.nodesource.com/gpgkey/nodesource-repo.gpg.key | sudo gpg --dearmor -o /etc/apt/keyrings/nodesource.gpg
+        echo "deb [signed-by=/etc/apt/keyrings/nodesource.gpg] https://deb.nodesource.com/node_$MIN_NODEJS.x nodistro main" | sudo tee /etc/apt/sources.list.d/nodesource.list
```
Comment (MarianRaphael) [inline service/raspbian-install-device-agent.sh:48]:
```suggestion
        echo "deb [signed-by=/etc/apt/keyrings/nodesource.gpg] https://deb.nodesource.com/node_18.x nodistro main" | sudo tee /etc/apt/sources.list.d/nodesource.list
```

File: service/raspbian-install-device-agent.sh (hunk 4 above)
Comment (MarianRaphael) [reply service/raspbian-install-device-agent.sh:58]: I will add it to the doc: https://flowfuse.com/docs/hardware/introduction/

File: service/raspbian-install-device-agent.sh (hunk 5 above)
Comment (MarianRaphael) [reply service/raspbian-install-device-agent.sh:109]: see [added sudo](https://github.com/FlowFuse/device-agent/pull/171/commits/0fb2eac669107100c6211ff69a0a3f0fb308a1fb)

File: service/raspbian-install-device-agent.sh (hunk 1 above)
Comment (hardillb) [reply service/raspbian-install-device-agent.sh:4]: Now all the commands are being run with `sudo` we shouldn't need  this warning, but we should swap the test to the negative and show a message saying it may ask for a password to run sudo only if not being run as root.

**Comment by hardillb** [comment]
@MarianRaphael Looks good, just needs the warning at the start updating to say it may ask for user password for sudo if not run as root.

**Comment by MarianRaphael** [comment]
@hardillb I updated the script and added the info text

**Review by hardillb — APPROVED** [review]
